docs/runtime-contract.md

```diff
@@ -285,10 +285,10 @@ container startup time (aka cold start time).
 On the initial deployment, platform providers SHOULD start an instance of the
 container to validate that the container is valid and will become ready. This
 startup SHOULD occur even if the container would not serve any user requests. If
-a container cannot satisfy the `livenessProbe` and `readinessProbe` during
-deployment startup, the deployment SHOULD be marked as failed.
+a container cannot satisfy the `readinessProbe` during deployment startup, the
+deployment SHOULD be marked as failed.
```

Comment:
Reading through #2972, it seems like the list given is a partial list of reasons that a Deployment may not become Ready. If the new containers are continually exiting due to failing Liveness probes, does that prevent e.g. Readiness probes from succeeding (maybe this only works if the Liveness duration is shorter than the Readiness duration)?

Reply:
> We should probably be explicit that the Revision SHOULD be marked as failed, since "the deployment" is a nebulous term here.

Thanks for pointing this out. I'll update the README after the other questions are sorted out. 

> If the new containers are continually exiting due to failing Liveness probes, does that prevent e.g. Readiness probes from succeeding (maybe this only works if the Liveness duration is shorter than the Readiness duration)?

Yes, I modified the probe configuration in `TestContainerExitingMsg` slightly so the readiness period is much longer than the liveness probe and the modified `TestContainerExitingMsg` test passed. 

```
...
       readinessProbe := &corev1.Probe{
		Handler: corev1.Handler{
			HTTPGet: &corev1.HTTPGetAction{},
		},
		PeriodSeconds: 1200,
	}
	livenessProbe := &corev1.Probe{
		Handler: corev1.Handler{
			HTTPGet: &corev1.HTTPGetAction{},
		},
		PeriodSeconds: 10,
	}
	if _, err := test.CreateConfiguration(t, clients, names, &test.Options{ReadinessProbe: readinessProbe, LivenessProbe: livenessProbe}); err != nil {
		t.Fatalf("Failed to create configuration %s: %v", names.Config, err)
	}
...
```

Complete test code commit in branch liveness-probe-conformance (ignore test image changes): https://github.com/yt3liu/serving/commit/fb3024cff6ee49645ea38d377fdb182fd958ab8f

Please let me know if this test doesn't answer your question and I will modify the test case to test the behavior mentioned. And do you see value in adding this case in the conformance test suite?

Reply:
Assuming that I've read through https://github.com/knative/serving/blob/master/test/conformance/errorcondition_test.go#L119 correctly, it sounds like the liveness probe doesn't directly affect readiness, but if it kills the container before the readiness probe succeeds, then the Revision won't become `Ready`.

Let me think about what expectations users might have in the case where they have attempted to deploy a container which fails the declared liveness check but passes a runtime check. What do you think the best experience would be? A few possible solutions:

1. Don't block `Ready`-ness, but the Revision containers will crash when the liveness probes fail, and there will be high container churn if traffic is migrated. 😢 
2. Require that Liveness checks have a shorter time-frame than Readiness checks so that the Revision fails to become `Ready` in the way described above. 😕 
3. Disallow Liveness checks in favor of only Readiness checks. 😥

Reply:
## Findings 

* the number of Pods never change. This includes the case where it passes the liveness, but not readiness. **Note:** I always see the message "Does not have minimum availability". Not sure if the autoscaler behaviour will be different if this is set. 

* when liveness fails, but readiness passes, the pod eventually reach CrashLoopBackOff after restarted 7 times (tested with test image readiness=(1), liveness=(3))

* when readiness fails, but liveness passes, it shows a warning "containers with unready status" and the pod never restarts (tested with test image readiness=(3), liveness=(1)) 

## Test Methods 

Created a test image to test the liveness, readiness by switching the http path it's checking https://github.com/yt3liu/serving/commit/52b516c8bcb05d1f7ab8d330f01a41287066408f. Basically it's a service with three endpoints that returns different http response for different path: 
(1) /healthy always returns 200 
(2) /unhealthy always return 500 
(3) /heathz returns 200 for the first 10 seconds after the service is started and 500 after that. 

From the test code, tested the cases where Liveness duration is shorter than the Readiness duration and vice versa. https://github.com/yt3liu/serving/commit/e212b1a2c5ee11015c066a10306b44d15cd3784c